CAIPs/caip-345.md

```diff
@@ -0,0 +1,104 @@
+---
+caip: CAIP-345
+title: Wallet Service URL property
+author: Chris Smith (@chris13524)
+discussions-to: https://github.com/ChainAgnostic/CAIPs/discussions/345
+status: Draft
+type: Standard
+created: 2025-02-17
+updated: 2025-02-17
+requires: 25
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the CAIP.-->
+This CAIP defines a mechanism for CAIP-25 wallets to advertize support for a "wallet service" which can handle requests for specific wallet RPC methods instead of apps sending these requests directly to the wallet app.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+A short (~200 word) description of the technical issue being addressed.
+
+## Motivation
+<!--The motivation is critical for CAIP. It should clearly explain why the state of the art is inadequate to address the problem that the CAIP solves. CAIP submissions without sufficient motivation may be rejected outright.-->
+In protocols such as WalletConnect it is bad UX to redirect the user to a wallet to handle requests that are non-actionable to them. Examples of this include:
+- `wallet_getCapabilities` - Defined in [EIP-5792](https://eips.ethereum.org/EIPS/eip-5792#wallet_getcapabilities)
+- `wallet_getCallsStatus` - Defined in [EIP-5792](https://eips.ethereum.org/EIPS/eip-5792#wallet_getcallsstatus-rpc-specification)
+- `wallet_getAssets` - Defined in [EIP-7811](https://eip.tools/eip/7811)
+
+By defining a way for wallets to send requests to an external URL, the requests can be satisfied without needing the wallet app to be open.
+
+## Specification
+<!--The technical specification should describe the standard in detail. The specification should be detailed enough to allow competing, interoperable implementations. -->
+
+Apps SHOULD use the wallet service when available for a method, instead of calling the wallet directly.
+
+Wallets SHOULD implement fallback handling for all wallet service methods.
```

Comment:
What is the relationship between methods listed in `walletService` and methods available according to CAIP-25? If a method is listed in `walletService` must it also be part of the CAIP-25 session?

Not requiring them to be present in the session could allow wallets to indicate support for a method ONLY on the wallet service, avoiding the wallet potentially needing to forward methods to the wallet service (if apps don't implement the wallet service but do implement the method).

Reply:
> For example if a provider pre-validates a method is valid before checking if it should be forwarded to the walletService.

It's not only providers that validate the methods; it's also the dapp/wallet that does that. E.g. when the method is received and is not listed in CAIP-25 it most likely would be defined as invalid.